apps/reaper/lib/reaper/xml/xml_stream/sax_handler.ex

```diff
@@ -0,0 +1,159 @@
+defmodule XMLStream.SaxHandler do
+  @moduledoc false
+
+  defmodule State do
+    @moduledoc false
+    @enforce_keys [:tag_path, :emitter]
+    defstruct stack: [], accumulate: false, tag_stack: [], tag_path: [], emitter: nil
+
+    def new(fields \\ %{}) do
+      struct!(__MODULE__, fields)
+    end
+
+    def update(%__MODULE__{}=state, fields) do
+      struct!(state, fields)
+    end
+
+    @spec pop(%State{}, :stack | :tag_stack | :tag_path) :: {any(), %State{}}
+    def pop(%__MODULE__{}=state, key) when key in [:stack, :tag_stack, :tag_path] do
+      [item | rest] = Map.fetch!(state, key)
+
+      {item, State.update(state, [{key, rest}])}
+    end
+
+    def pop_current_tag(%__MODULE__{}=state, tag_name) do
+      {{^tag_name, attributes, content}, state} = State.pop(state, :stack)
+
+      {{tag_name, attributes, Enum.reverse(content)}, state}
+    end
+
+    def at_path?(%__MODULE__{tag_stack: tag_stack, tag_path: tag_path}, tag_name) do
+      [tag_name | tag_stack] == tag_path
+    end
+
+    def ascend(%__MODULE__{}=state) do
+      State.pop(state, :tag_stack)
+    end
+
+    def descend(%__MODULE__{tag_stack: tag_stack}=state, tag_name) do
+      State.update(state, tag_stack: [tag_name | tag_stack])
+    end
```

Comment:
Am I crazy or should these names be rearranged? I would consider ascending a stack to be putting thing onto it and descending to be popping them off.

Reply:
I was naming them like a depth-first search. So as you enter tags in the document, you're descending into the tree, then ascending again as you encounter the end tags.